x-pack/plugins/alerting/server/alerts_client/lib/strip_framework_fields.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isArray, isEmpty, isFunction, fromPairs, isObject, isPlainObject, omit } from 'lodash';
+import { ALERT_REASON } from '@kbn/rule-data-utils';
+import { alertFieldMap } from '@kbn/alerts-as-data-utils';
+import { RuleAlertData } from '../../types';
+
+const allowedFrameworkFields: string[] = [ALERT_REASON];
+
+/**
+ * Remove framework fields from the alert payload reported by
+ * the rule type. Fields are considered framework fields if they are
+ * defined in the "alertFieldMap". Framework fields should only be
+ * set by the alerting framework during rule execution.
+ */
+export const stripFrameworkFields = <AlertData extends RuleAlertData>(
+  payload: AlertData
+): AlertData => {
+  const keysToStrip = Object.keys(alertFieldMap).filter(
+    (key: string) => !allowedFrameworkFields.includes(key)
+  );
+  // lodash omit can leave empty objects so we strip them from the result
+  return removeEmptyObjects(omit(payload, keysToStrip)) as AlertData;
+};
+
+type AlertTypes = string | number | boolean | object | AlertTypes[];
+type AlertRecord = Record<string, AlertTypes>;
+type AlertFields = AlertTypes | AlertRecord;
+
+const removeEmptyObjects = (data: AlertFields): AlertFields => {
```

Comment:
@mikecote Do you see any gotchas in this function? Using lodash `omit` leaves empty objects so we'd be left with fields like

```
kibana: {
  alert: {
    duration: {}
  }
}
```
if we had to strip `kibana.alert.duration.us` from the payload (for example), so I added this function but maybe it is overkill?

Reply:
Function looks good to me. To see if the function is necessary, I just did a check on the behaviours when calling the Elasticsearch index and update APIs to update an existing alert document. In both scenarios with empty objects, the call mutated the alert in the exact way as if it didn't have an empty object, so it seems not doing `removeEmptyObjects` will work the same.

When creating alerts, the created alert would just contain an empty object but I don't see any issues doing so either. So it seems safe to remove the `removeEmptyObjects` function and save an processing step.